### PR TITLE
Shared read-only pages between TAs (#2801 continued)

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This is #2801 rebased with a few fixes on top.

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
False positives from checkpatch

**Comment by jforissier** [comment]
> False positives from checkpatch

Hopefully fixed by https://github.com/OP-TEE/optee_os/pull/2974.

File: lib/libutils/isoc/bget_malloc.c [hunk 1]
```diff
@@ -181,6 +181,14 @@ static DEFINE_CTX(malloc_ctx);
 static __nex_data DEFINE_CTX(nex_malloc_ctx);
 #endif
 
+static void print_oom(size_t req_size __maybe_unused, void *ctx __maybe_unused)
+{
+#if defined(__KERNEL__) && defined(CFG_CORE_DUMP_OOM)
+	EMSG("mem alloc fail: sz %zu ctx %p", req_size, ctx);
```
Comment (jforissier) [inline lib/libutils/isoc/bget_malloc.c]: Minor: prefer "correct" syntax in `EMSG()` text. `"Memory allocation failed: size %zu context %p"`

**Comment by jforissier** [comment]
@jenswi-linaro thanks for rebasing and fixing the issues I stumbled upon with HiKey960. I hope we can merge this shortly now that 3.5.0 is out. We're only missing a few A-b/R-b, and I'm also going to run this through my usual test sequence that I use for OP-TEE releases.

**Comment by jenswi-linaro** [comment]
Fixed up "Introduce CFG_CORE_DUMP_OOM" and applied tags

**Comment by jenswi-linaro** [comment]
And hhank you @jforissier for the aggressive testing.
By the way I have patches in the pipe to replace "core: invalidate entire icache", but I think we'd better take those after this PR is merged.

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey960, GP)`

Tested at commit 004ba3f62dcc "core: share sections of loaded elf", but perhaps the T-b had better be applied to "core: share sections of loaded elf" and this very commit moved to last position?

**Comment by jenswi-linaro** [comment]
I guess you mean  004ba3f "core: user_ta_enter(): clean mapped params". Yes, it make sense to move "core: share sections of loaded elf" last since hopefully most of the errors fixed are triggered by that commit.

I'll reorder the commits and double check that the content is the same so you tag can be reused.

**Comment by jenswi-linaro** [comment]
Reordered to have e172be8b72fd ("core: share sections of loaded elf") last, TB added to that commit.

**Comment by jenswi-linaro** [comment]
There's a few commits that still needs review:
- `core: user_ta_enter(): clean mapped params`
- `core: fix assertion '(*pgt)->vabase == pg_info->va_base'`
- `core: tee_pager_set_uta_area_attr(): check same context`
- `core: tee_mmu.c: only free unused page tables`
- `core: tee_ta_init_user_ta_session(): flush pgt on error`
- `core: invalidate entire icache`
- `core: elf_flags_to_mattr(): add privileged bits`
- `core: tee_pager_set_uta_area_attr(): save flags`
- `core: add a file abstraction`


Next time I force push, I'll add my R-B to

- `Add TA entry point function: __ta_entry()`

**Comment by jenswi-linaro** [comment]
Found a bug (by inspection) in "core: add a file abstraction"

File: core/arch/arm/kernel/user_ta.c [hunk 2]
```diff
@@ -332,16 +332,16 @@ static TEE_Result user_ta_enter(TEE_ErrorOrigin *err,
 	/* Copy out value results */
 	update_from_utee_param(param, usr_params);
 
+	s = tee_ta_pop_current_session();
+	assert(s == session);
+cleanup_return:
```
Comment (jforissier) [inline core/arch/arm/kernel/user_ta.c]: I don't get it. This label is used only when `tee_mmu_map_param()` fails, in this case why is `tee_mmu_clean_param()` needed? Usually when a function fails, its normal actions do not need to be undone.

File: core/arch/arm/mm/core_mmu.c [hunk 3]
```diff
@@ -1354,18 +1354,30 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
...
+				*pgt = SLIST_NEXT(*pgt, link);
+				/* We should have alloced enough */
+				assert(*pgt);
+			}
 			assert((*pgt)->vabase == pg_info->va_base);
-#endif
+			pg_info->table = (*pgt)->tbl;
+#else
+			assert(*pgt); /* We should have alloced enough */
+			pg_info->table = (*pgt)->tbl;
```
Comment (jforissier) [inline core/arch/arm/mm/core_mmu.c]: Is it expected that we do `*pgt = SLIST_NEXT(*pgt, link);` after this line, but we don't in the first branch of the `#ifdef`?

File: core/arch/arm/kernel/user_ta.c (hunk 2 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/user_ta.c]: OK, I'll restore this function and add a cleanup call in `tee_mmu_map_param()`.

File: core/arch/arm/mm/core_mmu.c (hunk 3 above)
Comment (jenswi-linaro) [reply core/arch/arm/mm/core_mmu.c]: Yes, there's no `vabase` in the `pgt` without `CFG_PAGED_USER_TA` so the pgt are altways completely rewritten each time `core_mmu_populate_user_map()` is called. If we'd ever cache the filled in tables even without `CFG_PAGED_USER_TA` we would perhaps need to do something about that.

File: core/arch/arm/kernel/user_ta.c (hunk 2 above)
Comment (jforissier) [reply core/arch/arm/kernel/user_ta.c]: For "core: user_ta_enter(): clean mapped params" + the fixup commit:

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/arch/arm/mm/core_mmu.c (hunk 3 above)
Comment (jforissier) [reply core/arch/arm/mm/core_mmu.c]: OK, then for commit `core: fix assertion '(*pgt)->vabase == pg_info->va_base'`:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
> Revieweing core: tee_pager_set_uta_area_attr(): check same context.
>
>s/an are/an area/

I'll fix when force pushing.

>
>This sounds like an optimization ("it's enough") but a simple comparison is replaced by a loop... Is it a >bug fix instead? What was wrong with the previous assert?

It's a bug fix, the problem with the previous assert was that it assumed that only one area may be used.

**Comment by jforissier** [comment]
For commit "core: tee_pager_set_uta_area_attr(): check same context": I wouldn't mind you adding the word "fix" somewhere in the commit text, anyways:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
Applied fixups and tags. It looks like all commits now has a tag (not counting my s-o-b:s) now.
Note that "core: share sections of loaded elf" is the last commit and that it has Jerome's tested by tag.

**Comment by jforissier** [comment]
@jenswi-linaro awesome :) As a general principle, let's wait until CI is done, then I'll merge this PR.

**Comment by jforissier** [comment]
I think the IBART hang in `regression_1021` is caused by missing commits (the TA panic rework). Can you rebase onto master please?

**Comment by jenswi-linaro** [comment]
Rebased on master

**Comment by jforissier** [comment]
IBART is happy now, so this is ready to be merged. Thanks!

**Pull request merged**